core/src/main/scala/kafka/log/UnifiedLog.scala

```diff
@@ -629,10 +629,11 @@ class UnifiedLog(@volatile var logStartOffset: Long,
 
   /**
    * Return true if the given producer ID has a transaction ongoing.
+   * Note, if the incoming producer epoch is newer than the stored one, the transaction may have finished.
```

Comment:
nice! happy to see this added :) 
I think there is also a chance the epoch is less than the store one? So maybe in the comment we just say unequal? Or would we fence it before we even get to this point?

Reply:
At least from what I have read, we do not check the epoch before this point.

Reply:
I think it is possible and we should reject it earlier. We can make a separate PR for this optimization. https://issues.apache.org/jira/browse/KAFKA-18042